compiler/errors.go

```diff
@@ -0,0 +1,68 @@
+package compiler
+
+import (
+	"errors"
+	"fmt"
+)
+
+// ErrTooManyErrors is added to the ErrorList by the Trim method.
+var ErrTooManyErrors = errors.New("too many errors")
+
+// ErrorList wraps multiple errors as a single error.
+type ErrorList []error
+
+func (errs ErrorList) Error() string {
+	if len(errs) == 0 {
+		return "<no errors>"
+	}
+	return fmt.Sprintf("%s (and %d more errors)", errs[0].Error(), len(errs[1:]))
+}
+
+// ErrOrNil returns nil if ErrorList is empty, or the error otherwise.
+func (errs ErrorList) ErrOrNil() error {
+	if len(errs) == 0 {
+		return nil
+	}
+	return errs
+}
+
+// Append an error to the list.
+//
+// If err is an instance of ErrorList, the lists are concatenated together,
+// otherwise err is appended at the end of the list. If err is nil, the list is
+// returned unmodified.
+//
+// 	err := DoStuff()
+// 	errList := errList.Append(err)
+func (errs ErrorList) Append(err error) ErrorList {
+	if err == nil {
+		return errs
+	}
+	if err, ok := err.(ErrorList); ok {
```

Comment:
Not sure if we'll ever care about the 1.20 `Unwrap() []error` case here, too? Would be easy to add support if we want, and should be safe to add it "early", even before GopherJS is doing anything with 1.20.

On a related note, it would be easy to make `ErrorList` satisfy this interface... and whiel I can't imagine it would be needed, maybe semantically it would be expected? :shrug:

Reply:
I think supporting `Unwrap() []error` interface if fine, although I'm not sure if it would break `error.Is()` and `error.As()` in some weird way if compiled before 1.20?

Though even if we supported that interface, it don't think it would be semantically correct for us to unwrap multiple errors here, even if the incoming error theoretically implemented that interface. In general case the wrapping error may carry some useful information (like an error message), unwrapping would discard it here. `ErrorList` carries no semantics of its own other than "these errors happened", so flattening it is okay. Does that make sense?